src/openms/source/CHEMISTRY/SpectrumAnnotator.cpp

```diff
@@ -0,0 +1,390 @@
+
+    defaultsToParam_();
+  }
+
+  SpectrumAnnotator::SpectrumAnnotator(const SpectrumAnnotator & rhs) :
+    DefaultParamHandler(rhs)
+  {
+  }
+
+  SpectrumAnnotator & SpectrumAnnotator::operator=(const SpectrumAnnotator & rhs)
+  {
+    if (this != &rhs)
+    {
+      DefaultParamHandler::operator=(rhs);
+    }
+    return *this;
+  }
+
+  SpectrumAnnotator::~SpectrumAnnotator()
+  {
+  }
+
+  MSSpectrum<RichPeak1D> SpectrumAnnotator::annotateMatches(const PeptideHit& ph, const MSSpectrum<Peak1D>& spec, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    MSSpectrum<RichPeak1D> rich_theoretical_spec;
+    MSSpectrum<RichPeak1D> rich_input_spectrum = MSSpectrumHelper::clone(spec);
+
+    vector<pair<Size, Size> > al;
+    tg.getSpectrum(rich_theoretical_spec, ph.getSequence(), ph.getCharge()-1); //will get y5++, b2+, ... for precursor+++
+    if (!rich_theoretical_spec.isSorted())
+    {
+      rich_theoretical_spec.sortByPosition();
+    }
+    if (!rich_input_spectrum.isSorted())
+    {
+      rich_input_spectrum.sortByPosition();
+    }
+    sa.getSpectrumAlignment(al, rich_theoretical_spec, rich_input_spectrum); //peaks from theor. may be matched to none or one in spec!
+    for (vector<pair<Size, Size > >::const_iterator it = al.begin(); it != al.end(); ++it)
+    {
+        rich_input_spectrum[it->second].setMetaValue("IonMatchError", std::fabs(rich_input_spectrum[it->second].getMZ() - rich_theoretical_spec[it->first].getMZ()));
+        rich_input_spectrum[it->second].setMetaValue("IonName", rich_theoretical_spec[it->first].getMetaValue("IonName"));
+    }
+    Param sap = sa.getParameters();
+    rich_input_spectrum.setMetaValue("fragment_match_tolerance", sap.getValue("tolerance"));
+    return rich_input_spectrum;
+  }
+
+  void SpectrumAnnotator::addIonMatchStatistics(PeptideIdentification& pi, const MSSpectrum<Peak1D>& spec, const TheoreticalSpectrumGenerator& tg, const SpectrumAlignment& sa) const
+  {
+    if (!spec.empty())
+    {
+      for (vector<PeptideHit>::iterator ph = pi.getHits().begin(); ph != pi.getHits().end(); ++ph)
+      {
+        RichPeakSpectrum rich_spec = annotateMatches(*ph, spec, tg, sa);
+        rich_spec.sortByIntensity();
+
+        StringList ions;
+        double sum_intensity = 0;
+        double match_intensity = 0;
+        vector<double> fragmenterrors, intensities, mzs; // sorted by ascending intensity via rich_spec.sortByIntensity for topN statistics
+        fragmenterrors.reserve(rich_spec.size());
+        intensities.reserve(rich_spec.size());
+        mzs.reserve(rich_spec.size());
+
+        double nint = 0;
+        double cint = 0;
+
+        StringList allowed_types = ListUtils::create<String>("y,b,a,c,x,z");
+        map<String, vector<bool> > ion_series;
+        for (StringList::iterator st = allowed_types.begin(); st != allowed_types.end(); ++st)
+        {
+          ion_series.insert(make_pair(*st, vector<bool>(ph->getSequence().size()-1, false)));
```

Comment:
TSG it would not crash if e.g. a0 or a_size are generated but your code might - so maybe adding at leat a comment which ions are considered is a good idea for upcoming bug hunters ;)